connector/saml/saml.go

```diff
@@ -334,19 +353,97 @@ func (p *provider) HandlePOST(s connector.Scopes, samlResponse string) (ident co
 	return ident, nil
 }
 
+func (p *provider) validateStatus(resp *response) error {
+	status := resp.Status
+	if status == nil {
+		return fmt.Errorf("response did not contain a Status")
+	}
+	statusCode := status.StatusCode
+	if statusCode == nil {
+		return fmt.Errorf("response did not contain a StatusCode")
+	}
+	if statusCode.Value != "urn:oasis:names:tc:SAML:2.0:status:Success" {
+		return fmt.Errorf("status code of the response was not Success")
+	}
+	return nil
+}
+
+func (p *provider) validateSubjectConfirmation(subject *subject) error {
+	subjectConfirmation := subject.SubjectConfirmation
+	if subjectConfirmation == nil {
+		return fmt.Errorf("response did not contain a SubjectConfirmation")
+	}
+	subjectConfirmationData := subjectConfirmation.SubjectConfirmationData
+	if subjectConfirmationData == nil {
+		return fmt.Errorf("response did not contain a SubjectConfirmationData")
+	}
+	if subjectConfirmationData.Recipient != "" && subjectConfirmationData.Recipient != p.redirectURI {
+		return fmt.Errorf("expected recipient %q got %q", p.redirectURI, subjectConfirmationData.Recipient)
+	}
+	return nil
+}
+
+func (p *provider) validateConditions(assertion *assertion) error {
+	conditions := assertion.Conditions
+	if conditions == nil {
+		return fmt.Errorf("response did not contain a Conditions element")
+	}
+	now := p.now()
+	notBefore := time.Time(conditions.NotBefore)
+	if now.Add(p.allowedClockDrift).Before(notBefore) {
+		return fmt.Errorf("current time is before the NotBefore condition %v < %v", now.Add(p.allowedClockDrift), notBefore)
+	}
+	notAfter := time.Time(conditions.NotOnOrAfter)
+	if now.After(notAfter.Add(p.allowedClockDrift)) {
+		return fmt.Errorf("current time is after the NotOnOrAfter condition %v < %v", now, notAfter.Add(p.allowedClockDrift))
+	}
+	audienceRestriction := conditions.AudienceRestriction
+	if audienceRestriction == nil {
+		return fmt.Errorf("response did not contain an AudienceRestriction")
+	}
+	audiences := audienceRestriction.Audiences
+	if audiences == nil || len(audiences) == 0 {
+		return fmt.Errorf("response did not contain any Audience")
+	}
+	issuerInAudiences := false
+	for _, audience := range audiences {
+		if audience.Audience == p.issuer {
+			issuerInAudiences = true
+		}
+	}
+	if !issuerInAudiences {
+		return fmt.Errorf("%q is not a valid audience for this response", p.issuer)
+	}
+	return nil
+}
+
 // verify checks the signature info of a XML document and returns
 // the signed elements.
 func verify(validator *dsig.ValidationContext, data []byte) (signed []byte, err error) {
 	doc := etree.NewDocument()
-	if err := doc.ReadFromBytes(data); err != nil {
+	if err = doc.ReadFromBytes(data); err != nil {
 		return nil, fmt.Errorf("parse document: %v", err)
 	}
-
-	result, err := validator.Validate(doc.Root())
-	if err != nil {
-		return nil, err
+	verified := false
+	response := doc.Root()
+	transformedResponse, err := validator.Validate(response)
```

Comment:
In the old code only the signature of the document root / response element has been verfied. The `Validate` function of the goxmldsig library only looks for signatures on the root level. Therefore I have added the validation of the assertion element and replace it with the transformed one in case of success. The corresponding code in the python implementation is here https://github.com/onelogin/python-saml/blob/master/src/onelogin/saml2/response.py#L283-L305.